File: player/player.c
/*
 * Copyright (C) 2008, 2009, 2010, 2011 Philip Chimento and Marijn van Vliet.
 * All rights reserved.
 *
 * Chimara is free software copyrighted by Philip Chimento and Marijn van Vliet.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 3. Neither of the names Philip Chimento or Marijn van Vliet, nor the name of
 *    any other contributor may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <glib.h>
#include <glib-object.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <libchimara/chimara-glk.h>
#include <libchimara/chimara-if.h>
#include "player.h"
#include "app.h"
#include "error.h"
#include "util.h"

typedef struct _ChimaraPlayerPrivate {
	int dummy;
} ChimaraPlayerPrivate;

#define CHIMARA_PLAYER_PRIVATE(o) (G_TYPE_INSTANCE_GET_PRIVATE((o), CHIMARA_TYPE_PLAYER, ChimaraPlayerPrivate))
#define CHIMARA_PLAYER_USE_PRIVATE ChimaraPlayerPrivate *priv = CHIMARA_PLAYER_PRIVATE(self)

G_DEFINE_TYPE(ChimaraPlayer, chimara_player, GTK_TYPE_WINDOW);

static void
change_window_title(ChimaraGlk *glk, GParamSpec *pspec, GtkWindow *window)
{
	gchar *program_name, *story_name, *title;
	g_object_get(glk, "program-name", &program_name, "story-name", &story_name, NULL);
	if(!program_name) {
		gtk_window_set_title(window, "Chimara");
		return;
	}
	else if(!story_name)
		title = g_strdup_printf("%s - Chimara", program_name);
	else
		title = g_strdup_printf("%s - %s - Chimara", program_name, story_name);
	
	g_free(program_name);
	g_free(story_name);
	gtk_window_set_title(window, title);
	g_free(title);
}

static void
chimara_player_dispose(GObject *object)
{
	ChimaraPlayer *self = CHIMARA_PLAYER(object);
	if(chimara_glk_get_running(CHIMARA_GLK(self->glk))) {
		chimara_glk_stop(CHIMARA_GLK(self->glk));
		chimara_glk_wait(CHIMARA_GLK(self->glk));
	}
	
	/* Chain up */
	G_OBJECT_CLASS(chimara_player_parent_class)->dispose(object);
}

static void
chimara_player_finalize(GObject *object)
{
	g_object_unref(CHIMARA_PLAYER(object)->glk);
	
	/* Chain up */
	G_OBJECT_CLASS(chimara_player_parent_class)->finalize(object);
}

static void
chimara_player_class_init(ChimaraPlayerClass *klass)
{
	/* Override methods of parent classes */
	GObjectClass *object_class = G_OBJECT_CLASS(klass);
	object_class->dispose = chimara_player_dispose;
	object_class->finalize = chimara_player_finalize;

	/* Private data */
	g_type_class_add_private(klass, sizeof(ChimaraPlayerPrivate));
}

static void
chimara_player_init(ChimaraPlayer *self)
{	
	GError *error = NULL;
	ChimaraApp *theapp = chimara_app_get();

	/* Set parent properties */
	g_object_set(self,
		"title", _("Chimara"),
		"default-width", 600,
		"default-height", 800,
		NULL);

	/* Construct user interface */
	char *object_ids[] = {
		"actiongroup",
		"player-vbox",
		NULL
	};
	GtkBuilder *builder = new_builder_with_objects(object_ids);

	GtkActionGroup *actiongroup = GTK_ACTION_GROUP(load_object(builder, "actiongroup"));

	/* Set the default value of the "View/Toolbar" menu item upon creation of a
	 new window to the "show-toolbar-default" setting, but bind the setting
	 one-way only - we don't want toolbars to disappear suddenly */
	GtkToggleAction *toolbar_action = GTK_TOGGLE_ACTION(load_object(builder, "toolbar"));
	gtk_toggle_action_set_active(toolbar_action, g_settings_get_boolean(theapp->state_settings, "show-toolbar-default"));
	g_settings_bind(theapp->state_settings, "show-toolbar-default", toolbar_action, "active", G_SETTINGS_BIND_SET);

	self->glk = chimara_if_new();
	g_object_set(self->glk,
				 "ignore-errors", TRUE,
				 /*"interpreter-number", CHIMARA_IF_ZMACHINE_TANDY_COLOR,*/
				 NULL);
	char *default_css = get_data_file_path("style.css");
	if( !chimara_glk_set_css_from_file(CHIMARA_GLK(self->glk), default_css, &error) ) {
		error_dialog(GTK_WINDOW(self), error, "Couldn't open default CSS file: ");
	}
	
	/* DON'T UNCOMMENT THIS your eyes will burn
	 but it is a good test of programmatically altering just one style
	 chimara_glk_set_css_from_string(CHIMARA_GLK(glk),
	 "buffer.normal { font-family: 'Comic Sans MS'; }");*/

	GtkUIManager *uimanager = new_ui_manager("player.menus");
	gtk_ui_manager_insert_action_group(uimanager, actiongroup, 0);
	gtk_ui_manager_insert_action_group(uimanager, chimara_app_get_action_group(theapp), 1);
	GtkWidget *menubar = gtk_ui_manager_get_widget(uimanager, "/player_menu");
	self->toolbar = gtk_ui_manager_get_widget(uimanager, "/player_toolbar");
	gtk_widget_set_no_show_all(self->toolbar, TRUE);
	if(gtk_toggle_action_get_active(toolbar_action))
		gtk_widget_show(self->toolbar);
	else
		gtk_widget_hide(self->toolbar);
	
	/* Connect the accelerators */
	GtkAccelGroup *accels = gtk_ui_manager_get_accel_group(uimanager);
	gtk_window_add_accel_group(GTK_WINDOW(self), accels);

	GtkBox *vbox = GTK_BOX(load_object(builder, "player-vbox"));
	gtk_box_pack_end(vbox, self->glk, TRUE, TRUE, 0);
	g_object_ref(self->glk); /* add an extra reference to keep it alive while
							  the Glk program shuts down */
	gtk_box_pack_start(vbox, menubar, FALSE, FALSE, 0);
	gtk_box_pack_start(vbox, self->toolbar, FALSE, FALSE, 0);
	gtk_container_add(GTK_CONTAINER(self), GTK_WIDGET(vbox));
	
	gtk_builder_connect_signals(builder, self);
	g_signal_connect(self->glk, "notify::program-name", G_CALLBACK(change_window_title), self);
	g_signal_connect(self->glk, "notify::story-name", G_CALLBACK(change_window_title), self);

	g_object_unref(builder);
	g_object_unref(uimanager);
}

/* PUBLIC FUNCTIONS */

GtkWidget *
chimara_player_new(void)
{
    return GTK_WIDGET(g_object_new(CHIMARA_TYPE_PLAYER,
		"type", GTK_WINDOW_TOPLEVEL,
		NULL));
}

/* GLADE CALLBACKS */

#if 0
/* If a game is running in @glk, warn the user that they will quit the currently
running game if they open a new one. Returns TRUE if no game was running.
Returns FALSE if the user cancelled. Returns TRUE and shuts down the running
game if the user wishes to continue. */
static gboolean
confirm_open_new_game(ChimaraGlk *glk)
{
	g_return_val_if_fail(glk && CHIMARA_IS_GLK(glk), FALSE);
	
	GtkWindow *window = GTK_WINDOW(gtk_widget_get_toplevel(GTK_WIDGET(glk)));
	
	if(chimara_glk_get_running(glk)) {
		GtkWidget *dialog = gtk_message_dialog_new(window,
		    GTK_DIALOG_MODAL | GTK_DIALOG_DESTROY_WITH_PARENT,
		    GTK_MESSAGE_WARNING,
		    GTK_BUTTONS_CANCEL,
		    _("Are you sure you want to open a new game?"));
		gtk_message_dialog_format_secondary_text(GTK_MESSAGE_DIALOG(dialog),
		    _("If you open a new game, you will quit the one you are currently playing."));
		gtk_dialog_add_button(GTK_DIALOG(dialog), GTK_STOCK_OPEN, GTK_RESPONSE_OK);
		gint response = gtk_dialog_run(GTK_DIALOG(dialog));
		gtk_widget_destroy(dialog);
		
		if(response != GTK_RESPONSE_OK)
			return FALSE;

		chimara_glk_stop(glk);
		chimara_glk_wait(glk);
	}
	return TRUE;
}
#endif

void
on_stop_activate(GtkAction *action, ChimaraPlayer *player)
{
	chimara_glk_stop(CHIMARA_GLK(player->glk));
}

void
on_copy_activate(GtkAction *action, ChimaraPlayer *player)
{
	GtkWidget *focus = gtk_window_get_focus(GTK_WINDOW(player));
	/* Call "copy clipboard" on any widget that defines it */
	if(GTK_IS_LABEL(focus) || GTK_IS_ENTRY(focus) || GTK_IS_TEXT_VIEW(focus))
		g_signal_emit_by_name(focus, "copy-clipboard");
}

void
on_paste_activate(GtkAction *action, ChimaraPlayer *player)
{
	GtkWidget *focus = gtk_window_get_focus(GTK_WINDOW(player));
	/* Call "paste clipboard" on any widget that defines it */
	if(GTK_IS_ENTRY(focus) || GTK_IS_TEXT_VIEW(focus))
		g_signal_emit_by_name(focus, "paste-clipboard");
}

void
on_toolbar_toggled(GtkToggleAction *action, ChimaraPlayer *player)
{
	if(gtk_toggle_action_get_active(action))
		gtk_widget_show(player->toolbar);
	else
		gtk_widget_hide(player->toolbar);
}

void
on_undo_activate(GtkAction *action, ChimaraPlayer *player)
{
	chimara_glk_feed_line_input(CHIMARA_GLK(player->glk), "undo");
}

void 
on_save_activate(GtkAction *action, ChimaraPlayer *player)
{
	chimara_glk_feed_line_input(CHIMARA_GLK(player->glk), "save");
}

void 
on_restore_activate(GtkAction *action, ChimaraPlayer *player)
{
	chimara_glk_feed_line_input(CHIMARA_GLK(player->glk), "restore");
}

void 
on_restart_activate(GtkAction *action, ChimaraPlayer *player)
{
	chimara_glk_feed_line_input(CHIMARA_GLK(player->glk), "restart");
}

void 
on_quit_activate(GtkAction *action, ChimaraPlayer *player)
{
	chimara_glk_feed_line_input(CHIMARA_GLK(player->glk), "quit");
}

File: player/player.py
#!/usr/bin/env python

import sys
import os.path
import argparse
from gi.repository import Gdk
Gdk.threads_init()
from gi.repository import GObject, GLib, Gdk, Gio, Gtk, Chimara
import config

if config.ENABLE_NLS:
    import gettext
    gettext.install(config.GETTEXT_PACKAGE, config.PACKAGE_LOCALE_DIR,
        unicode=True, codeset='UTF-8')
else:
    _ = lambda x: x


class Player(GObject.GObject):
    __gtype_name__ = 'ChimaraPlayer'

    def __init__(self, graphics_file=None):
        super(Player, self).__init__()

        # Initialize settings file; it can be overridden by a "chimara-config"
        # file in the current directory
        if os.path.exists('chimara-config'):
            keyfile = 'chimara-config'
        else:
            keyfile = os.path.expanduser('~/.chimara/config')
        try:
            # This only works on my custom-built gobject-introspection; opened
            # bug #682702
            backend = Gio.keyfile_settings_backend_new(keyfile,
                "/org/chimara-if/player/", None)
        except AttributeError:
            backend = None
        self.prefs_settings = Gio.Settings('org.chimara-if.player.preferences',
            backend=backend)
        self.state_settings = Gio.Settings('org.chimara-if.player.state',
            backend=backend)

        builder = Gtk.Builder()
        try:
            builder.add_from_file(os.path.join(config.PACKAGE_DATA_DIR,
                'chimara.ui'))
        except GLib.GError:
            if config.DEBUG:
                builder.add_from_file(os.path.join(config.PACKAGE_SRC_DIR,
                    'chimara.ui'))
            else:
                raise
        self.window = builder.get_object('chimara')
        self.aboutwindow = builder.get_object('aboutwindow')
        self.prefswindow = builder.get_object('prefswindow')
        actiongroup = builder.get_object('actiongroup')

        # Set the default value of the "View/Toolbar" menu item upon creation
        # of a new window to the "show-toolbar-default" setting, but bind the
        # setting one-way only - we don't want toolbars to disappear suddenly
        toolbar_action = builder.get_object('toolbar')
        toolbar_action.active = \
            self.state_settings.get_boolean('show-toolbar-default')
        self.state_settings.bind('show-toolbar-default', toolbar_action,
            'active', Gio.SettingsBindFlags.SET)

        filt = Gtk.RecentFilter()
        # TODO: Use mimetypes and construct the filter dynamically depending on
        # what plugins are installed
        for pattern in ['*.z[1-8]', '*.[zg]lb', '*.[zg]blorb', '*.ulx', '*.blb',
            '*.blorb']:
            filt.add_pattern(pattern)
        recent = builder.get_object('recent')
        recent.add_filter(filt)

        uimanager = Gtk.UIManager()
        try:
            uimanager.add_ui_from_file(os.path.join(config.PACKAGE_DATA_DIR,
                'chimara.menus'))
        except GLib.GError:
            if config.DEBUG:
                uimanager.add_ui_from_file(os.path.join(config.PACKAGE_SRC_DIR,
                    'chimara.menus'))
            else:
                raise
        uimanager.insert_action_group(actiongroup)
        menubar = uimanager.get_widget('/menubar')
        toolbar = uimanager.get_widget('/toolbar')
        toolbar.no_show_all = True
        if toolbar_action.active:
            toolbar.show()
        else:
            toolbar.hide()

        # Connect the accelerators
        accels = uimanager.get_accel_group()
        self.window.add_accel_group(accels)

        self.glk = Chimara.IF(ignore_errors=True,
            # interpreter_number=Chimara.IFZmachineVersion.TANDY_COLOR,
            graphics_file=graphics_file)
        css_file = _maybe(self.prefs_settings.get_value('css-file'))
        if css_file is None:
            css_file = 'style.css'
        self.glk.set_css_from_file(css_file)

        # DON'T UNCOMMENT THIS your eyes will burn
        # but it is a good test of programmatically altering just one style
        # self.glk.set_css_from_string("buffer{font-family: 'Comic Sans MS';}")

        vbox = builder.get_object('vbox')
        vbox.pack_end(self.glk, True, True, 0)
        vbox.pack_start(menubar, False, False, 0)
        vbox.pack_start(toolbar, False, False, 0)

        builder.connect_signals(self)
        self.glk.connect('notify::program-name', self.change_window_title)
        self.glk.connect('notify::story-name', self.change_window_title)

        # Create preferences window
        # TODO

    def change_window_title(self, glk, pspec, data=None):
        if glk.props.program_name is None:
            title = "Chimara"
        elif glk.props.story_name is None:
            title = "{interp} - Chimara".format(interp=glk.props.program_name)
        else:
            title = "{interp} - {story} - Chimara".format(
                interp=glk.props.program_name,
                story=glk.props.story_name)
        self.window.props.title = title

    def on_open_activate(self, action, data=None):
        if not self.confirm_open_new_game():
            return

        dialog = Gtk.FileChooserDialog(_('Open Game'), self.window,
            Gtk.FileChooserAction.OPEN,
            (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL,
            Gtk.STOCK_OPEN, Gtk.ResponseType.ACCEPT))

        # Get last opened path
        path = _maybe(self.state_settings.get_value('last-open-path'))
        if path is not None:
            dialog.set_current_folder(path)

        response = dialog.run()
        dialog.hide()
        if response != Gtk.ResponseType.ACCEPT:
            return

        gamefile = dialog.get_file()
        self.search_for_graphics_file(gamefile.get_path())
        try:
            self.glk.run_game_file(gamefile)
        except GLib.Error as e:
            error_dialog(self.window, _('Could not open game file {filename}: {errmsg}').format(filename=gamefile.get_path(), errmsg=e.message))
            return

        path = dialog.get_current_folder()
        if path is not None:
            self.state_settings.last_open_path = path

        # Add file to recent files list
        manager = Gtk.RecentManager.get_default()
        uri = gamefile.get_uri()
        manager.add_item(uri)

        dialog.destroy()

    def on_recent_item_activated(self, chooser, data=None):
        if not self.confirm_open_new_game():
            return

        uri = chooser.get_current_uri()
        gamefile = Gio.file_new_for_uri(uri)

        self.search_for_graphics_file(gamefile.get_path())
        try:
            self.glk.run_game_file(gamefile)
        except GLib.Error as e:
            error_dialog(self.window,
                _('Could not open game file {filename}: {errmsg}').format(
                    filename=gamefile.get_basename(),
                    errmsg=e.message))
            return

        # Add file to recent files list again, this updates it to most recently
        # used
        manager = Gtk.RecentManager.get_default()
        manager.add_item(uri)

    def on_stop_activate(self, *args):
        self.glk.stop()

    def on_quit_chimara_activate(self, *args):
        Gtk.main_quit()

    def on_copy_activate(self, *args):
        focus = self.window.get_focus()
        # Call "copy clipboard" on any widget that defines it
        if (isinstance(focus, Gtk.Label)
            or isinstance(focus, Gtk.Entry)
            or isinstance(focus, Gtk.TextView)):
            focus.emit('copy-clipboard')

    def on_paste_activate(self, *args):
        focus = self.window.get_focus()
        # Call "paste clipboard" on any widget that defines it
        if isinstance(focus, Gtk.Entry) or isinstance(focus, Gtk.TextView):
            focus.emit('paste-clipboard')

    def on_preferences_activate(self, *args):
        self.prefswindow.present()

    def on_toolbar_toggled(self, action, *args):
        if action.get_active():
            self.toolbar.show()
        else:
            self.toolbar.hide()

    def on_undo_activate(self, *args):
        self.glk.feed_line_input('undo')

    def on_save_activate(self, *args):
        self.glk.feed_line_input('save')

    def on_restore_activate(self, *args):
        self.glk.feed_line_input('restore')

    def on_restart_activate(self, *args):
        self.glk.feed_line_input('restart')

    def on_quit_activate(self, *args):
        self.glk.feed_line_input('quit')

    def on_about_activate(self, *args):
        self.aboutwindow.set_version(config.PACKAGE_VERSION)
        self.aboutwindow.present()

    def on_window_delete_event(self, *args):
        Gtk.main_quit()
        return True

    def confirm_open_new_game(self):
        """
        If a game is running in the Glk widget, warn the user that they will
        quit the currently running game if they open a new one. Returns True if
        no game was running. Returns False if the user cancelled. Returns True
        and shuts down the running game if the user wishes to continue.
        """
        if not self.glk.props.running:
            return True

        dialog = Gtk.MessageDialog(self.window,
            Gtk.DialogFlags.MODAL | Gtk.DialogFlags.DESTROY_WITH_PARENT,
            Gtk.MessageType.WARNING, Gtk.ButtonsType.CANCEL,
            _("Are you sure you want to open a new game?"))
        dialog.format_secondary_text(
            _("If you open a new game, you will quit the one you are "
            "currently playing."))
        dialog.add_button(Gtk.STOCK_OPEN, Gtk.ResponseType.OK)
        response = dialog.run()
        dialog.hide()

        if response != Gtk.ResponseType.OK:
            return False

        self.glk.stop()
        self.glk.wait()
        return True

    def search_for_graphics_file(self, filename):
        """Internal function: See if there is a corresponding graphics file"""

        # First get the name of the story file
        base = os.path.basename(filename)
        base_noext = os.path.splitext(base)[0]

        # Check in the stored resource path, if set
        resource_path = _maybe(self.prefs_settings.get_value('resource-path'))

        # Otherwise check in the current directory
        if resource_path is None:
            resource_path = os.path.dirname(filename)

        blorbfile = os.path.join(resource_path, base_noext + '.blb')
        if os.path.exists(blorbfile):
            self.glk.graphics_file = blorbfile

    # Various signal handlers for GtkBuilder file
    def gtk_widget_hide(self, widget, *args):
        return Gtk.Widget.hide(widget)

    def gtk_widget_hide_on_delete(self, widget, *args):
        return Gtk.Widget.hide_on_delete(widget)

    def dummy_handler(self, *args):
        pass

    on_resource_file_set = dummy_handler
    on_interpreter_cell_changed = dummy_handler
    on_toggle_underline = dummy_handler
    on_toggle_italic = dummy_handler
    on_toggle_bold = dummy_handler
    on_toggle_justify = dummy_handler
    on_toggle_right = dummy_handler
    on_toggle_center = dummy_handler
    on_toggle_left = dummy_handler
    on_background_color_set = dummy_handler
    on_foreground_color_set = dummy_handler
    on_font_set = dummy_handler
    on_css_filechooser_file_set = dummy_handler


def _maybe(variant):
    """Gets a maybe value from a GVariant - not handled in PyGI"""
    v = variant.get_maybe()
    if v is None:
        return None
    return v.unpack()


def error_dialog(parent, message):
    dialog = Gtk.MessageDialog(parent, Gtk.DialogFlags.DESTROY_WITH_PARENT,
        Gtk.MessageType.ERROR, Gtk.ButtonsType.OK, message)
    dialog.run()
    dialog.destroy()

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('game_file', nargs='?', default=None,
        metavar='GAME FILE', help='the game file to load and start')
    parser.add_argument('graphics_file', nargs='?', default=None,
        metavar='GRAPHICS FILE', help='a Blorb resource file to include')
    args = parser.parse_args()

    # Create configuration dir ~/.chimara
    try:
        os.mkdir(os.path.expanduser('~/.chimara'))
    except OSError:
        # already exists
        assert os.path.isdir(os.path.expanduser('~/.chimara'))

    player = Player(graphics_file=args.graphics_file)
    player.window.show_all()

    if args.game_file is not None:
        try:
            player.glk.run_game(args.game_file)
        except GLib.Error as e:
            error_dialog(player.window,
                _("Error starting Glk library: {errmsg}").format(
                    errmsg=e.message))
            sys.exit(1)

    Gdk.threads_enter()
    Gtk.main()
    Gdk.threads_leave()

    player.glk.stop()
    player.glk.wait()

    sys.exit(0)
